### PR TITLE
[FLINK-19984][core] Add TypeSerializerTestCoverageTest to check whether tests based on SerializerTestBase and TypeSerializerUpgradeTestBase

### PR DESCRIPTION
**Pull request by SteNicholas**
## What is the purpose of the change

*Currently, `TypeInfoTestCoverageTest` which checks a test that extends `TypeInformationTestBase` for all type infos. But `TypeSerializer` doesn’t have the same thing that would verify that `TypeSerializer` has tests that extend `SerializerTestBase` and `TypeSerializerUpgradeTestBase`. Therefore this should add `TypeSerializerTestCoverageTest` to check whether to have tests based on `SerializerTestBase` and `TypeSerializerUpgradeTestBase` because all serializers should have tests based on both of them.*

## Brief change log

  - *Add `TypeSerializerTestCoverageTest` to check whether to have tests based on `SerializerTestBase` and `TypeSerializerUpgradeTestBase`.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (**yes** / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by guoweiM** [comment]
Thanks @SteNicholas  for opening this pr. In general the pr is good to me. I just have some minor comments.

1. Provide more precise filtering. Currently it is fine that we use the “typeutils” to filter because there are already some cases. (For example PrimitiveArraySerializerUpgradeTest.java and BasicTypeSerializerUpgradeTest.java). But I think the filter scope might be a little big. If somebody adds a serializer in the “typeutils” package and does not follow our rule we could not find it.
2. Make the filter consistent in the ‘TypeInfoTestCoverageTest’ and ‘TypeSerializerTestCoverageTest’. I think we might use the same condition to filter the “test/itcase/innercalss”.
3. Maybe we could split the commit to two. One is to add a test for StreamElement. One is to add the `‘TypeSerializerTestCoverageTest’`

**Comment by SteNicholas** [comment]
> Thanks @SteNicholas for opening this pr. In general the pr is good to me. I just have some minor comments.
> 
> 1. Provide more precise filtering. Currently it is fine that we use the “typeutils” to filter because there are already some cases. (For example PrimitiveArraySerializerUpgradeTest.java and BasicTypeSerializerUpgradeTest.java). But I think the filter scope might be a little big. If somebody adds a serializer in the “typeutils” package and does not follow our rule we could not find it.
> 2. Make the filter consistent in the ‘TypeInfoTestCoverageTest’ and ‘TypeSerializerTestCoverageTest’. I think we might use the same condition to filter the “test/itcase/innercalss”.
> 3. Maybe we could split the commit to two. One is to add a test for StreamElement. One is to add the `‘TypeSerializerTestCoverageTest’`

Thanks for @guoweiM detailed review. I have followed the second comment for the filter consistent.
About the first point, the typeutils package shoud be excluded. Because based on `SerializerTestBase `, the typeutils package has missing tests as follows:
```
Could not find test 'org.apache.flink.api.common.typeutils.base.GenericArraySerializerTest' that covers 'org.apache.flink.api.common.typeutils.base.GenericArraySerializer'.
Could not find test 'org.apache.flink.api.common.typeutils.SingleThreadAccessCheckingTypeSerializerTest' that covers 'org.apache.flink.api.common.typeutils.SingleThreadAccessCheckingTypeSerializer'.
Could not find test 'org.apache.flink.api.common.typeutils.UnloadableDummyTypeSerializerTest' that covers 'org.apache.flink.api.common.typeutils.UnloadableDummyTypeSerializer'.
Could not find test 'org.apache.flink.api.common.typeutils.base.VoidSerializerTest' that covers 'org.apache.flink.api.common.typeutils.base.VoidSerializer'.
Could not find test 'org.apache.flink.api.java.typeutils.runtime.CopyableValueSerializerTest' that covers 'org.apache.flink.api.java.typeutils.runtime.CopyableValueSerializer'.
Could not find test 'org.apache.flink.api.common.typeutils.base.NullValueSerializerTest' that covers 'org.apache.flink.api.common.typeutils.base.NullValueSerializer'.
``` 
About the third point, IMO, this could be included in this commit because the `TypeSerializerTestCoverageTest ` checkes these tests in typesutils package that the name of the tests are correct but not extend `SerializerTestBase`. Therefore, to pass the tests cases for `TypeSerializerTestCoverageTest`, these tests that the pull request included are necessary.

**Comment by SteNicholas** [comment]
@guoweiM , I have added the missing tests for `typeutils` package. Please help to review again if you are available.

**Comment by SteNicholas** [comment]
@guoweiM , according to offline discussion, I have add type serializer whitelist for missing `SerializerTestBase` and `TypeSerializerUpgradeTestBase`, which tests should be add later on other pull requests. This pull request is defined to find the missing test and check the other `TypeSerializer ` added in future.

File: flink-tests/src/test/java/org/apache/flink/test/completeness/TypeInfoTestCoverageTest.java [hunk 1]
```diff
@@ -50,10 +50,11 @@ public void testTypeInfoTestCoverage() {
...
 			if (Modifier.isAbstract(typeInfo.getModifiers()) ||
 					Modifier.isPrivate(typeInfo.getModifiers()) ||
-					typeInfo.getName().contains("Test$") ||
-					typeInfo.getName().contains("TestBase$") ||
-					typeInfo.getName().contains("ITCase$") ||
-					typeInfo.getName().contains("$$anon") ||
+					typeInfo.getName().contains("test") ||
+					typeInfo.getName().contains("Test") ||
+					typeInfo.getName().contains("ITCase") ||
+					typeInfo.getName().contains("$") ||
```
Comment (guoweiM) [inline flink-tests/src/test/java/org/apache/flink/test/completeness/TypeInfoTestCoverageTest.java:56]: I think we could not exclude all the internal classes.

**Comment by SteNicholas** [comment]
@guoweiM , according to offline discussion, I have add type serializer whitelist for missing `SerializerTestBase` and `TypeSerializerUpgradeTestBase`, which tests should be add later on other pull requests. This pull request is defined to find the missing test and in future this would add the missing `SerializerTestBase` and `TypeSerializerUpgradeTestBase` for the existed `TypeSerializer ` added .

**Comment by guoweiM** [comment]
Thanks @SteNicholas for resolving the comments and it looks good to me.

**Comment by SteNicholas** [comment]
@aljoscha ,  could you please take a look at this pull request? @guoweiM has already completed the review.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/KeyAndValueSerializer.java [hunk 2]
```diff
@@ -41,7 +41,7 @@
  *      [key-length] | &lt;key&gt; | &lt;timestamp&gt; | &lt;record&gt;
  * </pre>
  */
-final class KeyAndValueSerializer<IN> extends TypeSerializer<Tuple2<byte[], StreamRecord<IN>>> {
+public final class KeyAndValueSerializer<IN> extends TypeSerializer<Tuple2<byte[], StreamRecord<IN>>> {
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/KeyAndValueSerializer.java:44]: We shouldn't make this `public` just to make it accessible in the test. We can just use a `"String"` for exclusion.

File: flink-tests/src/test/java/org/apache/flink/test/completeness/TypeSerializerTestCoverageTest.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+		// check if a test exists for each type serializer
+		for (Class<? extends TypeSerializer> typeSerializer : typeSerializers) {
+			// we skip abstract classes, test classes, inner classes and scala classes to skip type serializer defined in test classes
+			if (Modifier.isAbstract(typeSerializer.getModifiers()) ||
+					Modifier.isPrivate(typeSerializer.getModifiers()) ||
+					typeSerializer.getName().contains("test") ||
+					typeSerializer.getName().contains("Test") ||
+					typeSerializer.getName().contains("ITCase") ||
+					typeSerializer.getName().contains("$$anon") ||
+					typeSerializer.getName().contains("scala")) {
```
Comment (aljoscha) [inline flink-tests/src/test/java/org/apache/flink/test/completeness/TypeSerializerTestCoverageTest.java:161]: These exclusions are a bit too broad now. Like this, newly added Scala `TypeSerializers` will not trigger the test.

**Review by aljoscha — APPROVED** [review]
The test looks very good overall! I had two comments that I will address myself before merging. 👍

**Comment by SteNicholas** [comment]
@aljoscha , please review the `TypeSerializerCoverageTest` again. I have merged your commit about the serializers in the Scala package.

**Comment by aljoscha** [comment]
Thanks for the contribution, @SteNicholas! And thanks for the review, @guoweiM! I now merged this.

**Pull request closed without merging**